Clarify that instrument name conflict only happens within a Meter (#2372)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

In order to avoid conflicts, views which specify a name SHOULD have an
instrument selector that selects at most one instrument. For the registration
mechanism described above, where selection is provided via configuration, the

> SDK SHOULD NOT allow Views with a specified name to be declared with instrument

Reviewer: This can't be enforced. When a view is declared, it is not known if it will match more than one instrument in the same meter.

Author: > This can't be enforced.

Here we picked the wording "SHOULD NOT" rather than "MUST NOT" so it's not enforcement.

> When a view is declared, it is not known if it will match more than one instrument in the same meter.

That's correct, here we are saying that folks should not allow wild card name or something that allows multiple matches.

Reviewer: Consider this situation:

``` python
meter_provider = MeterProvider(
    views=[View(instrument_name="foo.bar*", name="baz")],
)
meter1 = meter_provider.get_meter("library1")
meter2 = meter_provider.get_meter("library2")

meter1.create_counter("foo.bar")
meter2.create_counter("foo.bar")
```

Here we have a situation that will not cause conflicts in the output metrics (because the view will match instruments in different meters) but the wording of the spec will not allow it.

The case above can be easily fixed by the user by just removing the `*` in the instrument name, so maybe it is ok that the spec forbids a valid case like this one.

Author: Exactly, the intention is to not allow it for now. If later we want to allow it, we can (which is an additive change rather than breaking change).

Reviewer: All right, resolving.

> selectors that select more than one instrument (e.g. wild card instrument name)

Reviewer: ```suggestion
selectors that may select more than one instrument (e.g. wild card instrument name)
```
It is not known when the view is declared if it will match more than one instrument in the same meter or not.

> SDK SHOULD NOT allow Views with a specified name to be declared with instrument
> selectors that select more than one instrument (e.g. wild card instrument name)
> in the same Meter.

Reviewer: ```suggestion
A view with a specified name should match at most one instrument in the same Meter.
```

Reviewer: This should also specify which one of the instruments in the same meter that can match the view is the one that would be matched or indicate that it is not specified which one of the instruments in the same meter that can match the view is the one that would be matched.

Author: From the SDK perspective, all of them would match.
When it comes to particular exporter, things might be different - e.g. OTLP exporter (based on the data model) supports this scenario. For exporters that don't support this (e.g. Prometheus), we should spec it out in the exporter doc (e.g. that's why Prometheus Exporter spec is in its current state).

Reviewer: Ok, I think the intention of the spec is to ban the declaration of a named view with any instrument selector that may end up matching more than one instrument in the same meter _without specifying which instrument selectors can do so_. Changing my request for changes to comment.
